libr/asm/arch/8051/8051_disas.c

```diff
@@ -21,79 +21,131 @@ int _8051_disas (ut64 pc, RAsmOp *op, const ut8 *buf, ut64 len) {
+					sprintf (disasm, name, buf[1], buf[0] & mask);
+					val1 = buf[1];
 				} else if (arg1 == A_BIT) {
 					// bit addressing mode
-					sprintf (op->buf_asm, name, arg_bit (buf[1]), buf[1] & 0x07);
+					sprintf (disasm, name, arg_bit (buf[1]), buf[1] & 0x07);
+					val1 = buf[1];
 				} else {
 					// direct, immediate, bit
-					sprintf (op->buf_asm, name, buf[1]);
+					sprintf (disasm, name, buf[1]);
+					val1 = buf[1];
 				}
 			} else {
 				strcpy (op->buf_asm, "truncated");
+				free (disasm);
 				return -1;
 			}
 			break;
 		case 3:
 			if (len>2) {
 				if ((arg1 == A_ADDR16) || (arg1 == A_IMM16)) {
-					sprintf (op->buf_asm, name, 0x100 * buf[1] + buf[2]);
+					sprintf (disasm, name, 0x100 * buf[1] + buf[2]);
 				} else if (arg1 == A_IMM16) {
-					sprintf (op->buf_asm, name, 0x100 * buf[1] + buf[2]);
+					sprintf (disasm, name, 0x100 * buf[1] + buf[2]);
 				} else if (arg2 == A_OFFSET) {
 					if (mask != A_NONE) {
 						// @Ri, immediate, offset; Rn, immediate, offset
-						sprintf (op->buf_asm, name, buf[0] & mask, buf[1], arg_offset (pc + 3, buf[1]));
+						sprintf (disasm, name, buf[0] & mask, buf[1], arg_offset (pc + 3, buf[1]));
 					} else if (arg1 == A_BIT) {
 						// bit, offset
-						sprintf (op->buf_asm, name, arg_bit (buf[1]), buf[1] & 0x07, arg_offset (pc + 3, buf[2]));
+						sprintf (disasm, name, arg_bit (buf[1]), buf[1] & 0x07, arg_offset (pc + 3, buf[2]));
+						val1 = buf[1];
 					} else {
 						// direct, offset; a, immediate, offset
-						sprintf (op->buf_asm, name, buf[1], arg_offset (pc + 3, buf[2]));
+						sprintf (disasm, name, buf[1], arg_offset (pc + 3, buf[2]));
+						val1 = buf[1];
 					}
 				} else if (arg3 == A_OFFSET) {
 					// @Ri/Rn, direct, offset
-					sprintf (op->buf_asm, name, buf[0] & mask, buf[1], arg_offset (pc + 3, buf[2]));
+					sprintf (disasm, name, buf[0] & mask, buf[1], arg_offset (pc + 3, buf[2]));
+					val2 = buf[1];
 				} else if (arg1 == A_DIRECT && arg2 == A_DIRECT) {
 					// op direct, direct has src and dest swapped
-					sprintf (op->buf_asm, name, buf[2], buf[1]);
+					sprintf (disasm, name, buf[2], buf[1]);
+					val1 = buf[2];
+					val2 = buf[1];
 				} else {
 					// direct, immediate
-					sprintf (op->buf_asm, name, buf[1], buf[2]);
+					sprintf (disasm, name, buf[1], buf[2]);
+					val1 = buf[1];
 				}
 			} else {
 				strcpy (op->buf_asm, "truncated");
+				free (disasm);
 				return -1;
 			}
 			break;
 		default:
 			// if we get here something is wrong
+			free (disasm);
 			return 0;
 		}
+
+		// substitute direct addresses with register name
+		char key[10];
+		char subst[10];
+		if (arg1 == A_DIRECT && _8051_regs[val1]) {
+			sprintf (key, " 0x%02x", val1);
+			sprintf (subst, " %s", _8051_regs[val1]);
+			disasm = r_str_replace(disasm, key, subst, 0);
+		}
+		if (arg1 == A_BIT) {
+			val1 = arg_bit(val1);
+			if (_8051_regs[val1]) {
+				sprintf (key, "0x%02x.", val1);
```

Comment:
i would create a function that returns that disasm, by having val1 , because key and subst can be created in a single place instead of several times for each case. ths seems like a bit of repeated code to me